javascript-package/src/index.js

```diff
@@ -0,0 +1,460 @@
+import fetch from "isomorphic-fetch";
+import "core-js/stable";
+import "regenerator-runtime/runtime";
+import Fuse from "fuse.js";
+import { snakeCase } from "change-case";
+
+import pluralize from "./pluralize";
+import { BASE_URL, ENTITY_TYPES, LEAGUES, MIN_FUSE_SCORE } from "./constants";
+import {
+  validateLeagues,
+  validateStringArray,
+  validateUrlParameters,
+} from "./validators";
+import {
+  playersXgoalsParameters,
+  playersXpassParameters,
+  playersGoalsAddedParameters,
+  playersSalariesParameters,
+  goalkeepersXgoalsParameters,
+  goalkeepersGoalsAddedParameters,
+  teamsXgoalsParameters,
+  teamsXpassParameters,
+  teamsGoalsAddedParameters,
+  teamsSalariesParameters,
+  gamesXgoalsParameters,
+} from "./parameters.js";
+
+export default class Client {
+  #fuses = new Map();
+  #minimumFuseScore;
+
+  constructor({ minimumFuseScore } = {}) {
+    this.#minimumFuseScore = minimumFuseScore || MIN_FUSE_SCORE;
+  }
+
+  /* utils */
+  async #convertNameToId({ name, entityType, league }) {
+    const fuseKey = `${league}|${entityType}`;
+
+    let fuse;
+    if (!this.#fuses.has(entityType)) {
+      const pluralEntityType = pluralize(entityType);
+      const url = `${BASE_URL}${league}/${pluralEntityType}`;
+      const response = await fetch(url);
+
+      if (response.status >= 400) {
+        throw new Error(
+          `Got a bad response from the server: ${response.status}`
+        );
+      }
+
+      const entities = await response.json();
+      fuse = new Fuse(entities, {
+        includeScore: true,
+        keys: [`${entityType}_name`],
+      });
+      this.#fuses.set(fuseKey, fuse);
+    } else {
+      fuse = this.#fuses.get(fuseKey);
+    }
+
+    const result = fuse.search(name);
+
+    if (!result[0] || result[0].score > this.#minimumFuseScore) {
+      throw new Error(`Name ${name} does not match any known name`);
+    }
+
+    return result[0].item;
+  }
+
+  async #getEntityIdsByName({ names = [], entityType, leagues }) {
+    if (!names.length) {
+      const entities = new Map();
+
+      leagues.forEach((league) => {
+        entities.set(league, []);
+      });
+
+      return entities;
+    }
+
+    return Promise.all(
+      leagues.map(async (league) => {
+        const entities = await Promise.all(
+          names.map((name) =>
+            this.#convertNameToId({ name, entityType, league })
+          )
+        );
+
+        const entityIds = entities.map((entity) => entity[`${entityType}_id`]);
+
+        return { league, [`${pluralize(entityType)}`]: entityIds };
+      })
+    );
+  }
+
+  async #fetchEntity({ leagues, entityType, ids }) {
+    const results = await Promise.all(
+      leagues.map(async (league) => {
+        const url = `${BASE_URL}${league}/${pluralize(entityType)}${
+          ids.length ? `?${entityType}_id=${ids.join(",")}` : ""
+        }`;
```

Comment:
nitpick: might we want to use `URLSearchParams` for parsing this request?
thinking of something along the lines of:
```js
const params = new URLSearchParams();
if (ids.length) {
  params.append(`${entityType_id}`, ids.join(","));
}

const url = `${BASE_URL}${league}/${pluralize(entityType)}${params}`;
```

Reply:
I want to target both Node and the browser, and URLSearchParams aren't available in Node